File: harpy/lexer.py
from .token import Token
from .token_type import TokenType


class Lexer:
    """A very primitive lexer. Takes a string and splits it into a series of
    Tokens. Operators and punctuation are mapped to unique keywords. Names,
    which can be any series of letters, are turned into NAME tokens. All other
    characters are ignored (except to separate names). Numbers and strings are
    not supported. This is really just the bare minimum to give the parser
    something to work with.
    """

    _index: int
    _text: str
    _keywords: dict[str, TokenType]
    _simple_operators: dict[str, TokenType]
    _compound_operatos: dict[str, TokenType]

    def __init__(self, text: str):
        """Creates a new Lexer to tokenize the given string.

        Args:
            text (str): String to tokenize.
        """

        self._index = 0
        self._text = text
        self._keywords = {}
        self._compound_operators = {}
        self._simple_operators = {}

        for type in TokenType:
            keyword = type.keyword()
            if keyword is not None:
                self._keywords[keyword] = type
                continue

            compound_operator = type.compound_operator()
            if compound_operator is not None:
                self._compound_operators[compound_operator] = type
                continue

            simple_operator = type.simple_operator()
            if simple_operator is not None:
                self._simple_operators[simple_operator] = type
                continue

    def __iter__(self):
        return self

    def __next__(self):
        while self._index < len(self._text):
            c = self._text[self._index]
            self._index += 1

            match c:
                case "/":
                    # Potential comments.
                    match self._peek():
                        case "/":
                            return self._read_line_comment()
                        case "*":
                            return self._read_block_comment()
                        case _:
                            return Token(self._simple_operators[c], c)
                case '"' | "'":
                    return self._read_str_literal(c)
                case ".":
                    if self._peek().isalpha():
                        return self._read_bool_literal_or_logical()
                    else:
                        return self._read_num_literal(c)
                case "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9":
                    return self._read_num_literal(c)
                case _:
                    if (kw := self._read_keyword(c)) is not None:
                        return kw
                    elif (op := self._read_logical(c)) is not None:
                        return op
                    elif c + (c1 := self._peek()) in self._compound_operators:
                        self._advance()
                        return Token(self._compound_operators[c + c1], c + c1)
                    elif c in self._simple_operators:
                        return Token(self._simple_operators[c], c)
                    elif c.isalpha():
                        return self._read_name()
                    else:
                        # Ignore all other characters (whitespace, etc.)
                        pass

        # Once we've reached the end of the string, just return EOF tokens. We'll
        # just keep returning them as many times as we're asked so that the
        # parser's lookahead doesn't have to worry about running out of tokens.
        return Token(TokenType.EOF, "\0")

    def _read_line_comment(self) -> Token:
        start = self._index - 1

        # Consume second '/'.
        self._advance()

        while True:
            match self._peek():
                case "\n" | "\r" | "\0":
                    return Token(
                        TokenType.LINE_COMMENT, self._text[start : self._index]
                    )
                case _:
                    self._advance()

    def _read_block_comment(self) -> Token:
        start = self._index - 1

        while True:
            match self._advance():
                case "*":
                    match self._advance():
                        case "/":
                            return Token(
                                TokenType.BLOCK_COMMENT, self._text[start : self._index]
                            )
                        case "\0":
                            raise SyntaxError("Unterminated block comment.")
                        case _:
                            pass  # Do nothing, keep advancing.
                case "\0":
                    raise SyntaxError("Unterminated block comment.")
                case _:
                    pass  # Do nothing, keep advancing.

    def _read_bool_literal_or_logical(self) -> Token:
        literal = "."

        while (c := self._advance()) not in (".", "\0"):
            literal += c

        literal += "."

        match literal.lower():
            case ".t." | ".f.":
                return Token(TokenType.BOOL_LITERAL, literal)
            case ".or.":
                return Token(TokenType.OR, literal)
            case ".and.":
                return Token(TokenType.AND, literal)
            case _:
                raise SyntaxError(f"Unable to read token '{literal}'.")

    def _read_num_literal(self, c: str) -> Token:
        literal = c
        dotfound = False
        hexnum = False

        while not (c := self._peek()).isspace():
            match c.lower():
                case "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9":
                    literal += c
                case "x":
                    if literal == "0":
                        literal += c
                        hexnum = True
                    else:
                        raise SyntaxError(f"Unterminated hexadecimal literal '{literal + c}'.")
                case "a" | "b" | "c" | "d" | "e" | "f":
                    if hexnum:
                        literal += c
                    else:
                        raise SyntaxError(f"Invalid numeric literal '{literal + c}'.")
                case ".":
                    if not dotfound:
                        literal += c
                        dotfound = True
                    else:
                        raise SyntaxError(f"Second decimal point found in literal '{literal + c}'.")
                case "\0":
                    return Token(TokenType.NUM_LITERAL, literal)

            self._advance()

        return Token(TokenType.NUM_LITERAL, literal)

    def _read_str_literal(self, c: str) -> Token:
        literal = c
        endquote = c

        while (c := self._peek()) != endquote:
            match c:
                case "\0":
                    if len(literal) > 1 and literal[-1:] == endquote:
                        return Token(TokenType.STR_LITERAL, literal + self._advance())

                    raise SyntaxError(f"Unterminated string literal '{literal}'.")
                case _:
                    literal += c

            self._advance()

        return Token(TokenType.STR_LITERAL, literal + self._advance())

    def _read_keyword(self, kw: str) -> Token | None:
        start_index = self._index

        while (c := self._advance()).isalpha():
            kw += c

        if kw.lower() in self._keywords:
            return Token(self._keywords[kw.lower()], kw)

        self._index = start_index
        return None

    def _read_logical(self, op: str) -> Token | None:
        if op != ".":
            return None

        while (c := self._advance()) != ".":
            op += c
        op += c  # Final '.'.

        if op.lower() not in self._compound_operators:
            raise SyntaxError(f"Unterminated logical operator '{op}'.")

        return Token(self._compound_operators[op.lower()], op)

    def _read_name(self) -> Token:
        start = self._index - 1
        while self._index < len(self._text):
            if not self._text[self._index].isalpha():
                break
            self._index += 1

        name = self._text[start : self._index]
        return Token(TokenType.NAME, name)

    def _peek(self) -> str:
        if len(self._text) > self._index:
            return self._text[self._index]
        else:
            return "\0"

    def _advance(self) -> str:
        c = "\0"

        if len(self._text) > self._index:
            c = self._text[self._index]
            if len(self._text) > self._index:
                self._index += 1

        return c

File: src/lexer.rs
use std::collections::HashMap;

use crate::token::{Token, TokenType};

/// A very primitive lexer. Takes a string and splits it into a series of
/// Tokens. Operators and punctuation are mapped to unique keywords. Names,
/// which can be any series of letters, are turned into NAME tokens. All other
/// characters are ignored (except to separate names). Numbers and strings are
/// not supported. This is really just the bare minimum to give the parser
/// something to work with.
pub struct Lexer {
    index: usize,
    text: String,
    simple_operators: HashMap<String, TokenType>,
    compound_operators: HashMap<String, TokenType>,
}

impl Lexer {
    pub fn new(text: &str) -> Lexer {
        let mut lexer = Lexer {
            index: 0,
            text: String::from(text),
            simple_operators: HashMap::new(),
            compound_operators: HashMap::new(),
        };

        for t in TokenType::iterator() {
            let compound_operator = t.compound_operator();
            if compound_operator != "" {
                lexer
                    .compound_operators
                    .insert(String::from(compound_operator), t.clone());
            }
        }

        for t in TokenType::iterator() {
            let simple_operator = t.simple_operator();
            if simple_operator != "" {
                lexer
                    .simple_operators
                    .insert(String::from(simple_operator), t.clone());
            }
        }

        lexer
    }
}

impl Iterator for Lexer {
    type Item = Token;

    fn next(&mut self) -> Option<Token> {
        while self.index < self.text.len() {
            let c = self.text[self.index..].chars().next().unwrap();
            self.index += 1;

            match c {
                '/' => {
                    let token = match self.peek() {
                        '/' => Some(self.read_line_comment()),
                        '*' => Some(self.read_block_comment()),
                        _ => {
                            let charstr = format!("{}", c);
                            let op = self.simple_operators.get(&charstr);
                            match op {
                                Some(t) => {
                                    return Some(Token {
                                        token_type: *t,
                                        value: charstr,
                                    });
                                }
                                None => return None,
                            };
                        }
                    };

                    return token;
                }
                _ => {
                    let c1 = self.peek();
                    let compound_charstr = format!("{}{}", c, c1);
                    let compound_op = self.compound_operators.get(&compound_charstr);

                    let token = match compound_op {
                        Some(t) => {
                            let c2 = self.text[self.index..].chars().next().unwrap();
                            if c2 != '\0' {
                                self.index += 1;
                            }
                            return Some(Token {
                                token_type: *t,
                                value: compound_charstr,
                            });
                        }
                        None => {
                            let simple_charstr = format!("{}", c);
                            let simple_op = self.simple_operators.get(&simple_charstr);

                            let t = match simple_op {
                                Some(t1) => Some(Token {
                                    token_type: *t1,
                                    value: simple_charstr,
                                }),
                                None => {
                                    if c.is_alphabetic() {
                                        return Some(self.read_name());
                                    } else {
                                        continue;
                                    }
                                }
                            };

                            t
                        }
                    };

                    return token;
                }
            }
        }

        // Once we've reached the end of the string, just return EOF tokens. We'll
        // just keep returning them as many times as we're asked so that the
        // parser's lookahead doesn't have to worry about running out of tokens.
        Some(Token {
            token_type: TokenType::Eof,
            value: String::from("\0"),
        })
    }
}

impl Lexer {
    fn read_line_comment(&mut self) -> Token {
        let mut comment = String::from("//");

        // Consume second '/'.
        self.advance();

        loop {
            match self.peek() {
                '\n' | '\r' | '\0' => {
                    return Token {
                        token_type: TokenType::LineComment,
                        value: comment,
                    };
                }
                c => {
                    comment.push(c);
                    self.advance();
                }
            }
        }
    }

    fn read_block_comment(&mut self) -> Token {
        let mut comment = String::from("/*");

        // Consume '*'.
        self.advance();

        loop {
            let c = self.advance();
            match c {
                '*' => {
                    let c1 = self.advance();
                    comment.push(c);

                    match c1 {
                        '/' => {
                            comment.push(c1);
                            return Token {
                                token_type: TokenType::BlockComment,
                                value: comment,
                            };
                        }
                        '\0' => panic!("Unterminated block comment."),
                        _ => comment.push(c1),
                    }
                }
                '\0' => panic!("Unterminated block comment."),
                _ => comment.push(c),
            }
        }
    }

    fn read_name(&mut self) -> Token {
        let mut name = String::from("");

        for c in String::from(self.text.clone())[self.index - 1..].chars() {
            if c.is_alphabetic() {
                name.push(c);
                let c1 = self.peek();
                if c1.is_alphabetic() {
                    self.index += 1;
                }
            } else {
                return Token {
                    token_type: TokenType::Name,
                    value: name,
                };
            }
        }

        Token {
            token_type: TokenType::Name,
            value: name,
        }
    }

    fn peek(&mut self) -> char {
        let c = self.text[self.index..].chars().next();

        return match c {
            Some(c) => c,
            None => '\0',
        };
    }

    fn advance(&mut self) -> char {
        let c = '\0';

        match self.text[self.index..].chars().next() {
            Some(c) => {
                self.index += 1;
                return c;
            }
            None => (),
        }

        c
    }
}

#[cfg(test)]
mod tests {
    use super::{Lexer, Token, TokenType};

    #[test]
    pub fn test_next() {
        let obs = get_obs("from + offset(time)");

        let expected = vec![
            Token {
                token_type: TokenType::Name,
                value: String::from("from"),
            },
            Token {
                token_type: TokenType::Plus,
                value: String::from("+"),
            },
            Token {
                token_type: TokenType::Name,
                value: String::from("offset"),
            },
            Token {
                token_type: TokenType::LeftParen,
                value: String::from("("),
            },
            Token {
                token_type: TokenType::Name,
                value: String::from("time"),
            },
            Token {
                token_type: TokenType::RightParen,
                value: String::from(")"),
            },
        ];

        assert_eq!(format!("{:?}", obs), format!("{:?}", expected));
    }

    #[test]
    pub fn test_assign_vs_comma() {
        let mut obs = get_obs("a := b");

        let mut expected = vec![
            Token {
                token_type: TokenType::Name,
                value: String::from("a"),
            },
            Token {
                token_type: TokenType::Assign,
                value: String::from(":="),
            },
            Token {
                token_type: TokenType::Name,
                value: String::from("b"),
            },
        ];

        assert_eq!(format!("{:?}", obs), format!("{:?}", expected));

        obs = get_obs("a:b");

        expected = vec![
            Token {
                token_type: TokenType::Name,
                value: String::from("a"),
            },
            Token {
                token_type: TokenType::Colon,
                value: String::from(":"),
            },
            Token {
                token_type: TokenType::Name,
                value: String::from("b"),
            },
        ];

        assert_eq!(format!("{:?}", obs), format!("{:?}", expected));
    }

    #[test]
    pub fn test_relations() {
        let mut obs = get_obs("a == b");

        let mut expected = vec![
            Token {
                token_type: TokenType::Name,
                value: String::from("a"),
            },
            Token {
                token_type: TokenType::Eq1,
                value: String::from("=="),
            },
            Token {
                token_type: TokenType::Name,
                value: String::from("b"),
            },
        ];

        assert_eq!(format!("{:?}", obs), format!("{:?}", expected));

        obs = get_obs("a <= b");

        expected = vec![
            Token {
                token_type: TokenType::Name,
                value: String::from("a"),
            },
            Token {
                token_type: TokenType::Le,
                value: String::from("<="),
            },
            Token {
                token_type: TokenType::Name,
                value: String::from("b"),
            },
        ];

        assert_eq!(format!("{:?}", obs), format!("{:?}", expected));
    }

    #[test]
    pub fn test_line_comment() {
        let obs = get_obs("// This is a line comment.");

        let expected = vec![Token {
            token_type: TokenType::LineComment,
            value: String::from("// This is a line comment."),
        }];

        assert_eq!(format!("{:?}", obs), format!("{:?}", expected));
    }

    #[test]
    pub fn test_block_comment() {
        let mut obs = get_obs("/* This is a block comment.*/");

        let mut expected = vec![Token {
            token_type: TokenType::BlockComment,
            value: String::from("/* This is a block comment.*/"),
        }];

        assert_eq!(format!("{:?}", obs), format!("{:?}", expected));

        obs = get_obs(
            "
/* This is also a
 * block
 * comment.
 */
 ",
        );

        expected = vec![Token {
            token_type: TokenType::BlockComment,
            value: String::from("/* This is also a\n * block\n * comment.\n */"),
        }];

        assert_eq!(format!("{:?}", obs), format!("{:?}", expected));
    }

    #[test]
    #[should_panic(expected = "Unterminated block comment.")]
    fn test_block_comment_panics() {
        let _ = get_obs("/* This is an unfinished block comment.*");
    }

    #[test]
    #[should_panic(expected = "Unterminated block comment.")]
    fn test_block_comment_panics2() {
        let _ = get_obs("/* This one is even worse./*");
    }

    fn get_obs(source: &str) -> Vec<Token> {
        let lexer = Lexer::new(source);
        let mut obs = vec![];

        for token in lexer {
            if token.token_type == TokenType::Eof {
                break;
            }
            obs.push(token);
        }

        obs
    }
}
